library/core/src/slice/mod.rs

```diff
@@ -110,6 +110,27 @@ enum Direction {
     Back,
 }
 
+macro_rules! compile_assert_non_zero_array {
+    ($func_name:ident,$array_part_name:literal, $t:ty, $array_size:expr) => {{
+        trait CompileAssert {
+            const TRIGGER: ();
+        }
+        impl<T, const N: usize> CompileAssert for [T; N] {
+            const TRIGGER: () = if N == 0 {
+                panic!(concat!(
+                    $array_part_name,
+                    " size in `",
+                    stringify!($func_name),
+                    "` must be bigger than zero."
+                ))
+            } else {
+            };
+        }
+
+        let _ = <[$t; $array_size] as CompileAssert>::TRIGGER;
+    }};
+}
+
```

Comment:
As @nbdd0121 suggested, `const { assert!(N != 0); };` is a much simpler way to achieve this. It does still only result in a post-monomorphization error, but it's much cleaner.

Plus, inline const expressions are in the stabilization process right now: #104087

Reply:
Thank you for suggestion, now it looks much nicer.